File: packages/dart/lib/src/sentence/sentence_length_range.dart
import 'package:randino/src/constants.dart';
import 'package:randino/src/internal/utils.dart';
import 'package:randino/src/sentence/sentence_generator.dart';
import 'package:randino/src/types.dart';
import 'package:randino/src/word/data/index.dart';

/// Every sentence length the language can produce, in characters, which is what
/// `randSentence` falls back to when `minLength` or `maxLength` is omitted.
///
/// The lower end is the shortest shape with the shortest words in it, and the
/// upper end the longest shape with a modifier on every phrase, so the range is
/// wide on purpose — the shape of each sentence is picked inside it. A null
/// [language] spans every language at once.
///
/// ```dart
/// sentenceLengthRange(WordLanguage.ko); // LengthRange(6, 41)
/// sentenceLengthRange(WordLanguage.en); // LengthRange(13, 92)
/// ```
LengthRange sentenceLengthRange([WordLanguage? language]) {
  final languages = language == null ? wordLanguages : <WordLanguage>[language];
  var min = 1 << 30;
  var max = 0;

  for (final code in languages) {
    final range = naturalRange(code);

    if (range.min < min) min = range.min;
    if (range.max > max) max = range.max;
  }

  return LengthRange(
    clampInt(min, randLengthMin, randSentenceLengthMax),
    clampInt(max, randLengthMin, randSentenceLengthMax),
  );
}

File: packages/javascript/lib/sentence/sentenceLengthRange.ts
import { clamp } from '../_internal/utils.js';
import { RAND_LENGTH_MIN, RAND_SENTENCE_LENGTH_MAX } from '../constants.js';
import type { WordLanguageOption } from '../_types/global.js';
import { WORD_LANGUAGES } from '../word/data/index.js';
import { naturalRange } from './sentenceGenerator.js';

/**
 * Every sentence length the language can produce, in characters, which is what
 * `randSentence` falls back to when `minLength` or `maxLength` is omitted. The
 * lower end is the shortest shape with the shortest words in it, and the upper
 * end the longest shape with a modifier on every phrase, so the range is wide on
 * purpose — the shape of each sentence is picked inside it.
 *
 * @example
 * sentenceLengthRange('ko'); // [5, 45]
 * sentenceLengthRange('en'); // [10, 84]
 */
export function sentenceLengthRange(language: WordLanguageOption = 'all'): [number, number] {
	const languages = language === 'all' ? WORD_LANGUAGES : [language];
	let min = Infinity;
	let max = 0;

	for (const code of languages) {
		const [low, high] = naturalRange(code);

		min = Math.min(min, low);
		max = Math.max(max, high);
	}

	return [
		clamp(min, RAND_LENGTH_MIN, RAND_SENTENCE_LENGTH_MAX),
		clamp(max, RAND_LENGTH_MIN, RAND_SENTENCE_LENGTH_MAX)
	];
}
